File: TW-FINCH/run_on_dataset.m
function Result = run_on_dataset(Dataset, tw_finch, datasets_path)
% Input
% Dataset: dataset name - one of 'BF', 'Hollywood_extended', 'FS', 'MPII_Cooking', 'YTI'
% tw_finch: true or false: pass true to run tw_finch else will run finch
% datasets_path : root folder where all datasets folders exist
% Output
% Result: struct with computed metrics on full Dataset and a detailed overview table of reslts per video

%%%%%%%%%%%%%
if nargin < 3
    datasets_path='./Action_Segmentation_Datasets/';
end



if tw_finch
    fprintf('Running: %s on %s Dataset\n', 'TW-FINCH', Dataset)
else
    fprintf('Running: %s on %s Dataset\n', 'FINCH', Dataset)
end

 
feat_path=fullfile(datasets_path, Dataset, 'features');

gt_path=fullfile(datasets_path, Dataset, 'groundTruth');
mapping_path = fullfile(datasets_path, Dataset, 'mapping');

imds = imageDatastore(feat_path, 'IncludeSubfolders',true,'FileExtensions','.txt','LabelSource','foldernames');


files=imds.Files;

activity_video_labels=imds.Labels;

if strcmp(Dataset, 'MPII_Cooking')
    act={};
        for i=1:numel(files)
            buff = textscan(files{i},'%s','Delimiter','/');
            buff=buff{1}; 
            vid_name = buff{end}; vid_name=vid_name(1:end-4);
        n =textscan(vid_name, '%s', 'Delimiter', '-');
        act = [act; n{1}{2}];
        end
        activity_video_labels=categorical(act);
end
%%% load MPI test set Only %%%
MPI_flag = true;
 if strcmp(Dataset, 'MPII_Cooking') && MPI_flag
    [files, activity_video_labels] = get_MPI_testset(files, activity_video_labels, mapping_path);
 end
 
 
% Load average gt per activity for clustering each video 
activity_gt = load(fullfile(mapping_path, 'avg_gt_per_activity.mat'));
gt_per_activity = activity_gt.gt_per_activity; 
 
 
[moF,iou,f1, overview_table] = action_seg(files, activity_video_labels, gt_per_activity, Dataset, datasets_path, tw_finch);
 
Result.moF = moF;
Result.IoU = iou;
Result.F1 = f1;
Result.overview_table = overview_table;


fprintf('[INFO] : Results\n')
fprintf('DataSet: %s .... MoF:= %2.1f%% .... IoU:= %2.1f%%\n', Dataset, moF*100, iou*100 )

%% get MPI_ test set
    function [files, activity_labels] = get_MPI_testset(filenames, labs, mapping_path)        
        MPI_test_path=fullfile(mapping_path, 'sequencesTest.txt');
        tb= table2cell(readtable(MPI_test_path, 'ReadVariableNames', false));
        idx=[];
        for i=1:numel(tb);
            idx= [idx; find(contains(filenames, tb{i}))];
        end
        files = filenames(idx);
        activity_labels=labs(idx);       
    end
    
end

File: TW-FINCH/run_on_dataset.py
import glob
import os
import time
import numpy as np
import pandas as pd
from python.read_utils import get_mapping, read_gt_label, decode_hung_labels, estimate_cost_matrix, avg_gt_activity_datasets
from scipy.optimize import linear_sum_assignment
from sklearn import metrics
from python.twfinch import FINCH
import argparse

def run_twfinch(
        dataset_name='50Salads',
        datasets_path='./Action_Segmentation_Datasets',
        tw_finch=True,
        verbose=False
):
    # setup paths
    ds_name = dataset_name
    path_ds = os.path.join(datasets_path, ds_name)
    path_gt = os.path.join(path_ds, 'groundTruth/')
    path_mapping = os.path.join(path_ds, 'mapping', 'mapping.txt')
    path_features = os.path.join(path_ds, 'features/')

    # %% Load all needed files: Descriptor, GT & Mapping
    if os.path.exists(path_mapping):
        # Create the Mapping dict
        mapping_dict = get_mapping(path_mapping)
    else:
        mapping_dict = None

    # Load all filenames from ds path        
    if ds_name == "MPII_Cooking":  # Load MPI test set files
        mpi_df = pd.read_csv(os.path.join(path_ds, 'mapping', 'sequencesTest.txt'), names=['filename'], header=None)
        mpi_df['filename'] = path_features + mpi_df['filename'] + '-cam-002.txt'
        filenames = mpi_df.filename.tolist()
    else:
        filenames = glob.glob(os.path.join(path_features, '**/*.txt'), recursive=True)

    # Create the result matrix to hold all values in the end
    results_matrix_ds = np.zeros(shape=(len(filenames), 7), dtype=object)

    # %% Loop over each file
    for file_num, cur_filename in enumerate(filenames):

        # Load the Descriptor        
        cur_desc = np.loadtxt(cur_filename, dtype='float32')

        # Load the GT_labels, map them to the corresponding ID    
        video_name = os.path.basename(cur_filename)[:-4]
        if ds_name == "MPII_Cooking":
            activity_name = video_name.split("-")[1]
        else:
            activity_name = os.path.basename(os.path.split(cur_filename)[0])

        # n_clusters for TWFINCH
        n_clusters = int(avg_gt_activity_datasets[ds_name][activity_name])
        gt_label_path = os.path.join(path_gt, video_name)
        if not os.path.exists(gt_label_path):
            gt_label_path = os.path.join(path_gt, video_name + '.txt')

        gt_labels, n_labels = read_gt_label(gt_label_path, mapping_dict=mapping_dict)

        # cluster data with Finch        
        start = time.time()
        c, num_clust, req_c = FINCH(cur_desc, req_clust=n_clusters, verbose=verbose, tw_finch=tw_finch)

        # Find best assignment through Hungarian Method
        cost_matrix = estimate_cost_matrix(gt_labels, req_c)
        row_ind, col_ind = linear_sum_assignment(cost_matrix)
        # decode the predicted labels
        y_pred = decode_hung_labels(req_c, col_ind)

        # Calculate the metrics (External libraries)
        mof = -cost_matrix[row_ind, col_ind].sum() / len(
            cur_desc)  # MoF accuracy can also be computed with sklearn metrics.accuracy_score
        cur_acc = metrics.accuracy_score(gt_labels, y_pred)

        f1_macro = metrics.f1_score(gt_labels, y_pred, average='macro')  # F1-Score
        # iou_macro = metrics.jaccard_score(gt_labels, y_pred, average='macro')  # IOU
        # penalize equally over/under clustering
        iou = np.sum(metrics.jaccard_score(gt_labels, y_pred, average=None)) / n_clusters
        end = time.time()
        if verbose:
            print(f'Evaluation on Video {activity_name}/{video_name} finshed in {np.round(end - start)} seconds: accuracy = {cur_acc} IoU = {iou} and f1 ={f1_macro}')

        # Transfer all the calculated metrics into the result matrix 
        results_matrix_ds[file_num][0] = activity_name + '/' + video_name
        results_matrix_ds[file_num][1] = cur_acc  # Mean over frame accuracy
        results_matrix_ds[file_num][2] = iou  # IOU
        results_matrix_ds[file_num][3] = f1_macro  # F1 from Sklearn
        results_matrix_ds[file_num][4] = n_clusters  # Number of desired clusters for TW-Finch
        results_matrix_ds[file_num][5] = n_labels  # Number of ground truth clusters
        results_matrix_ds[file_num][5] = gt_labels  # Encoded GT Data
        results_matrix_ds[file_num][6] = y_pred  # Encoded cluster labels n_clusters     

    avg_score = np.mean(results_matrix_ds[:, 1:4], axis=0)
    print(f'Overal Results on {ds_name} Dataset: MOF:{avg_score[0]}, IOU: {avg_score[1]}, F-Score: {avg_score[2]}')
    return results_matrix_ds


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--dataset-name', required=True, help='Specify the path to your data csv file.')
    parser.add_argument('--datasets-path', required=True, help='Specify the root folder of all datsets')
    parser.add_argument('--tw-finch', action='store_true', default=True)
    parser.add_argument('--verbose', action='store_true', default=True)
    args = parser.parse_args()
    results = run_twfinch(dataset_name=args.dataset_name,
                          datasets_path=args.datasets_path,
                          tw_finch=args.tw_finch,
                          verbose=args.verbose)